### PR TITLE
Add counter example

### PR DESCRIPTION
**Pull request by bonbud-macryg**

### CODE REVIEW

File: pkg/arvo/neo/cod/std/src/imp/counter.hoon [hunk 1]
```diff
@@ -17,8 +17,10 @@
   ^-  (set stud:neo)
   (sy %counter-diff ~)
 ::
-::  counter can have no children
-::  beneath it in the tree
+::
+::  counter does not "constrain" its children;
```
Comment (hanfel-dovned) [inline pkg/arvo/neo/cod/std/src/imp/counter.hoon]: > counter does not "constrain" its children;

> counter does not constrain the type of its children

File: pkg/arvo/neo/cod/std/src/imp/counter.hoon [hunk 2]
```diff
@@ -36,11 +38,23 @@
     ::
     ::  de-vase the state; we don't know what it is,
     ::  in most cases it will be counter's old state
-    +*  this  !<(number state-vase)
+    +*  state  !<(number state-vase)
```
Comment (hanfel-dovned) [inline pkg/arvo/neo/cod/std/src/imp/counter.hoon]: let's get rid of the `+*` here and do it with a `=/` in the on-poke arm instead.

Comment (hanfel-dovned) [reply pkg/arvo/neo/cod/std/src/imp/counter.hoon]: and then collapse the stud and vase that get passed in along with the bowl into a pail, and assert that the stud is a %number. then we do know what the vase is :)

**Comment by hanfel-dovned** [comment]
This branch contains:
- Backend code for counter
- Frontend code for counter, compatible with the version of Sky in `wh/revert-role-conversions`
- Backend and frontend docs for counter in the `fil` directory

File: pkg/arvo/neo/cod/std/src/fil/counter.md [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+* The path `/~sampel/examples/counter/1` will be a pith `~[[%p ~sampel] %examples %counter [%ud 1]]`.
+
+A `note` is one of the four types of command any shrub will accept.
+
+```hoon
++$  note
+  $%  [%make made]             ::  create a shrub
+      [%poke =pail]            ::  poke a shrub
+      [%tomb cas=(unit case)]  ::  tombstone a shrub
+      [%cull ~]                ::  ???
```
Comment (bonbud-macryg) [inline pkg/arvo/neo/cod/std/src/fil/counter.md:232]: What does %cull do?

File: pkg/arvo/neo/cod/std/src/fil/counter.md [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+      ^-  (quip card:neo pail:neo)
+      ?>  =(%counter-diff stud)
+      =/  act  !<(counter-diff vaz)
+      ?>  =(-.act %inc)  ::  XX can we remove this line?
+      `number/!>(+(state))
+  --
+--
+```
+
+Let’s set up a fakeship that has `/base/app/neo.hoon`, the prototype “shrub runner”, then go over the code piece by piece.
```
Comment (hanfel-dovned) [inline pkg/arvo/neo/cod/std/src/fil/counter.md:107]: Dependent on this devex we decide on next week

**Comment by hanfel-dovned** [comment]
A lot of the devex/sky/folder structure stuff in this tutorial will be dependent on the devex we decide on. Maybe we'll split this out into another tutorial or leave it out all together and just explain the code semantic vs. how-to

**Comment by hanfel-dovned** [comment]
Closing this an merging into `wh/revert-role-conversions` instead.

**Pull request closed without merging**